### PR TITLE
Arguments everywhere

### PR DESCRIPTION
**Pull request by carmocca**
Fixes https://github.com/Lightning-AI/lit-gpt/issues/716

The objective of this PR is to get rid of globals across the repository.

The current design plan is to keep flat arguments at the CLI level that get converted into dataclasses. The dataclasses are used across different functions.

I created a tag https://github.com/Lightning-AI/lit-gpt/tree/globals pointing to the last commit before this PR

### CODE REVIEW

File: pretrain/tinyllama.py [hunk 1]
```diff
@@ -22,51 +22,41 @@
...
-beta2 = 0.95
-grad_clip = 1.0
-decay_lr = True
-min_lr = 4e-5
-
-batch_size = global_batch_size // devices
-gradient_accumulation_iters = batch_size // micro_batch_size
-assert gradient_accumulation_iters > 0
-warmup_iters = warmup_steps * gradient_accumulation_iters
-log_iter_interval = log_step_interval * gradient_accumulation_iters
```
Comment (awaelchli) [inline pretrain/tinyllama.py:62]: Where did this part go? It looks like these computations have simply disappeared.

File: pretrain/tinyllama.py [hunk 2]
```diff
@@ -333,6 +371,23 @@ def choose_logger(logger_name: str, name: str, resume: Union[bool, Path], *args,
...
+    unsupported = [
+        (io_args, ["train_data_dir", "val_data_dir", "checkpoint_dir"]),
+        (train_args, ["epoch_size", "epochs"]),
+        (eval_args, ["max_new_tokens"]),
+    ]
+    for args, names in unsupported:
+        for name in names:
+            if getattr(args, name) is not None:
+                raise ValueError(f"{__file__} doesn't support the {name!r} argument. This is set in {args}")
+    required = [(train_args, ["max_tokens", "max_norm"])]
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pretrain/tinyllama.py:384]: Because the user could pass `--max_tokens null` through command line or from a different config

File: pretrain/tinyllama.py [hunk 3]
```diff
@@ -333,6 +371,23 @@ def choose_logger(logger_name: str, name: str, resume: Union[bool, Path], *args,
...
+    for args, names in required:
+        for name in names:
+            if getattr(args, name) is None:
+                raise ValueError(f"{__file__} requires the {name!r} argument. This is set in {args}")
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pretrain/tinyllama.py:388]: Yes. What's the problem with that? Does it matter?

File: pretrain/tinyllama.py (hunk 1 above)
Comment (carmocca) [reply pretrain/tinyllama.py:62]: batch size and gradient accumulation iters are now dataclasses methods (see args.py)

log_iter_interval is in L182: https://github.com/Lightning-AI/lit-gpt/pull/913/files#diff-00fedfb359238295e8b95bc7c9f90975463bf4c40f03b2a010a118ec21f2f8d3R182

**Comment by rasbt** [comment]
Not sure where my current problem is, but even if I'm running

```
CUDA_VISIVLE_DEVICES=3 python dpo-tinyllama-experiments/lit-gpt/finetune/full.py ... --devices 1
```

it uses the cuda device 0 on a machine with 4 CUDA devices
